### ISSUE TITLE
Retrieval by instance name does not work

### ISSUE THREAD

**Issue by dalewking** [fluttercommunity/get_it#35]
Consider this simple example:
```
void main() {
  try {
    var sl = GetIt.instance;

    sl.registerSingleton("some String", instanceName: "foo");

    final x = sl<String>("foo");
  } catch (e) {
    print(e);
  }
}
```

This fails with this output:

> String
> Invalid argument(s): Object with name foo has a different type (String) than the one that is inferred (String) where you call it

I think the problem is this code:

        if (registeredObject.registrationType is! T) {
          print(T.toString());
          throw ArgumentError(
              "Object with name $instanceName has a different type (${registeredObject.registrationType.toString()}) than the one that is inferred (${T.toString()}) where you call it");
        }
 
registrationType has a type of `Type` so this code is asking whether an instance of `Type` has the type `String` and the answer is of course no. I think you want to replace `is!` with `!=`.

**Comment by dalewking** [comment]
Alternatively you might want to go ahead and get the object from the factory then see if the object returned from the factory is T which would let you retrieve a named instance using a super class type as in:

    sl.registerSingleton(someSubclassInstance, instanceName: "foo");

    SuperClassType x = sl("foo");

**Comment by knaeckeKami** [comment]
Yep, this is already fixed in master, see:
https://github.com/fluttercommunity/get_it/pull/32

As a workaround until a new release is shipped, you can just replace the type parameters with 
<dynamic>.

**Comment by dalewking** [comment]
@knaeckeKami Unfortunately, I think markdown syntax swallowed what you were trying to say.

**Comment by knaeckeKami** [comment]
My bad, edited.

**Comment by escamoteur** [comment]
Should be fixed with new version V3.1.0.
